kernel/aster-nix/src/ipc/semaphore/system_v/sem.rs

```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::vec::Vec;
+use core::time::Duration;
+
+use ostd::sync::{Mutex, Waiter, Waker};
+
+use super::sem_set::SEMVMX;
+use crate::{
+    ipc::{key_t, semaphore::system_v::sem_set::sem_sets, IpcFlags},
+    prelude::*,
+    process::{Pid, Process},
+    time::{clocks::JIFFIES_TIMER_MANAGER, timer::Timeout, SystemTime},
+};
+
+#[derive(Clone, Copy, Debug, Pod)]
+#[repr(C)]
+pub struct SemBuf {
+    sem_num: u16,
+    sem_op: i16,
+    sem_flags: i16,
+}
+
+impl SemBuf {
+    pub fn sem_num(&self) -> u16 {
+        self.sem_num
+    }
+
+    pub fn sem_op(&self) -> i16 {
+        self.sem_op
+    }
+
+    pub fn sem_flags(&self) -> i16 {
+        self.sem_flags
+    }
+}
+
+#[repr(u16)]
+#[derive(Debug)]
+enum Status {
+    Normal = 0,
+    Pending = 1,
+    Removed = 2,
+}
+
+struct PendingOp {
+    sem_buf: SemBuf,
+    status: Arc<Mutex<Status>>,
+    waker: Arc<Waker>,
+    pid: Pid,
+    process: Weak<Process>,
+}
+
+impl Debug for PendingOp {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("PendingOp")
+            .field("sem_buf", &self.sem_buf)
+            .field("status", &(*self.status.lock()))
+            .field("pid", &self.pid)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct Semaphore {
+    val: Mutex<i32>,
+    /// PID of the process that last modified semaphore.
+    /// - through semop with op != 0
+    /// - through semctl with SETVAL and SETALL
+    /// - through SEM_UNDO when task exit
+    latest_modified_pid: RwMutex<Pid>,
+    /// Pending alter operations. For each pending operation, it has `sem_op < 0`.
+    pending_alters: Mutex<Vec<PendingOp>>,
+    /// Pending zeros operations. For each pending operation, it has `sem_op = 0`.
+    pending_const: Mutex<Vec<PendingOp>>,
```

Comment:
Perhaps considering the use of `LinkedList` would be a better option in this case. The operations predominantly involve adding and removing items from the array, and it's worth noting that Linux also utilizes LinkedList. You have the option to employ LinkedList from `intrusive_collection`. 

It's uncertain how much of a performance advantage it would yield, so you can also postpone this decision until you attempt to enhance the performance.

Reply:
I have tested the use of `Vec` and `LinkedList`, and the results are 1.0080 and 1.0523 separately. It seems like the use of `Vec` is better. 

Maybe is because the testing will simply add and remove several semaphores and in this situation, `LinkedList` is required to allocate and free memory while the `Vec` doesn't need this operation.

Reply:
Putting the choice between `LinkedList` and `Vec` aside, I don't think it is a good idea to use `Vec<T>` when the size of `T` is far greater than a pointer, which is our case.

```rust
struct PendingOp {
    sem_buf: SemBuf,
    status: Arc<Mutex<Status>>,
    waker: Arc<Waker>,
    pid: Pid,
    process: Weak<Process>,
}
```

Inserting or removing elements from `Vec<T>` involves copying a great number of bytes when the size of `T` is large. So `Vec<Box<T>>` should be better.

If you have to use `Box` anyway, then `LinkedList` is better than `Vec`.